Allow configuration of locales

nickbrowne opened the pull request:
Eleanor and I just paired on this new feature to allow configuration of the `check_variables` task so that now you can specify which locales should be checked.

Like so:

```ruby
I18n::Hygiene::Tasks::CheckVariables.new do |task|
  task.check_locales = [:fr, :id, :es]
end
```

There's a decent amount of potential for refactoring (as evidenced by how difficult it was to test!), but we can leave that for later.

nickbrowne committed:
Allow configuration of locales for check_variables task

nickbrowne committed:
Improve logging

nickbrowne committed:
Add pry-nav as dev dependency

It sure is useful!

nickbrowne committed:
Deprecate unconfigurable check_variables rake task

nickbrowne commented on i18n-hygiene.gemspec:
```diff
@@ -11,5 +11,7 @@ Gem::Specification.new do |s|
 
   s.add_dependency 'i18n', ['~> 0.6', '>= 0.6.9']
   s.add_dependency 'parallel', '~> 1.3'
+  s.add_dependency 'rake', '~> 11.0.0'
```
I think this dependency on rake is a bit too strict, will fix

yob commented on lib/i18n/hygiene/variable_checker.rb:
```diff
@@ -5,15 +5,17 @@ module Hygiene
+      DEFAULT_LOCALES_TO_CHECK = [ :fr ]
 
-      def initialize(key, i18n_wrapper)
+      def initialize(key, i18n_wrapper, locales = DEFAULT_LOCALES_TO_CHECK)
```
Do we need to support the legacy case as a default here? Maybe the legacy `i18n:hygiene:check_variables` rake task could pass in `[:fr]`?

nickbrowne replied:
Yeah, I was thinking exactly that on the tram this morning

yob commented on lib/tasks/i18n_hygiene.rake:
```diff
@@ -26,6 +26,13 @@ namespace :i18n do
 
     desc "check for mismatching interpolation variables"
     task check_variables: :environment do
+      ActiveSupport::Deprecation.warn("i18n:hygiene:check_variables is deprecated", caller)
```
Can we safely use `ActiveSupport` if its not a dependency of this gem?

nickbrowne replied:
Ah indeed we cannot

yob reviewed (commented):
I know this isn't ready yet but I was excited about us making time for this work so I took it for a spin :)

With the following addition to a rake file in donations, it worked!

```ruby
require 'i18n/hygiene/tasks/check_variables'
I18n::Hygiene::Tasks::CheckVariables.new(foo: :environment) do |task|
  task.check_locales = [:fr, :id, :es]
end
```

then:
```bash
$ be rake spec:foo                           
Checking for mismatching interpolation variables...                        
rake aborted!                        
I18n::InvalidLocale: :id is not a valid locale
```

Ignore the `I18n::InvalidLocale` error - that's just because I asked the gem to check the `id` locale which doesn't exist on donations yet.

One thing worth noting is that I had to manully specify the `:environment` task as a dependency to ensure rails (and our locale data) was loaded.

> There's a decent amount of potential for refactoring (as evidenced by how difficult it was to test!), but we can leave that for later.

I agree that `I18n::Hygiene::VariableChecker` could use some refactor. Specifically, I found they way it used two public methods hard to wrap my head around. Maybe we could find a way to condense the API into a single method?

In any case, I agree thats a job for another PR. We could change that API without it affecting users of the library.

yob commented:
Oh, I like the style of configuration you've chosen too 👍

nickbrowne committed:
Just pass in :fr to deprecated rake task

Keeps the existing behaviour without having to maintain a constant
in the class.

nickbrowne committed:
Make rake dependency more permissive

nickbrowne commented:
I think this should be good to review again, cc @keithpitty

nickbrowne commented:
One thing probably worth mentioning is that originally we'd hoped to provide a unified class to create rake tasks from, but that ended up being quite complicated due to inheriting from `TaskLib` (which I'd also tried to avoid, but that's another story). 

In the end we decided it was better to have a  separate class for each task. I suspect once we do the work to create classes for each rake task, we might want to refactor.

nickbrowne commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[i18n-hygiene] Decide on a strategy for configuring gem](https://trello.com/c/Rfh3PY3N/7640-i18n-hygiene-decide-on-a-strategy-for-configuring-gem)

keithpitty commented on lib/i18n/hygiene/variable_checker.rb:
```diff
@@ -4,16 +4,14 @@ module Hygiene
+      def initialize(key, i18n_wrapper, locales = [])
         @key = key
         @i18n_wrapper = i18n_wrapper
+        @locales = locales
```
Should the elements of `locales` be validated?

nickbrowne replied:
Potentially, but should we do that in this branch? This is PR mostly concerned with setting up a pattern to allow configuration, which unblocks the other cards we've got queued.

keithpitty replied:
Fair call. I'll buy that.

nickbrowne replied:
:ok_hand:

nickbrowne commented:
> One thing worth noting is that I had to manully specify the :environment task as a dependency to ensure rails (and our locale data) was loaded.

Can you think of a way to avoid this? It does seem like a bit of a gotcha.

nickbrowne commented:
> Can you think of a way to avoid this? It does seem like a bit of a gotcha.

To add to this, Elearnor and I discovered an issue in analytics where the JS translations appeared not to be covered by the hygiene checks. I'll add another card for that.

keithpitty commented:
> I agree that I18n::Hygiene::VariableChecker could use some refactor. Specifically, I found they way it used two public methods hard to wrap my head around. Maybe we could find a way to condense the API into a single method?

I've added a [card](https://trello.com/c/yugGxvi2/7679-i18n-hygiene-refactor-to-improve-internal-structure-and-public-api) to remind us to do this.

keithpitty commented:
@nickbrowne in which projects do we currently use this gem?

nickbrowne commented:
Donations and Analytics, I believe

keithpitty commented:
> Donations and Analytics, I believe

Have you tested that this branch works as expected with Donations and Analytics?

yob commented on i18n-hygiene.gemspec:
```diff
@@ -11,5 +11,7 @@ Gem::Specification.new do |s|
 
   s.add_dependency 'i18n', ['~> 0.6', '>= 0.6.9']
   s.add_dependency 'parallel', '~> 1.3'
+  s.add_dependency 'rake', '>= 0.8.7'
```
Very permissive!

nickbrowne replied:
It's what railties uses :)

nickbrowne commented:
I have, but in doing so discovered that it wasn't really working as it is because we don't seem to be checking JS translations. There's a card for that here https://trello.com/c/H6TpX5Ve/7678-i18n-hygiene-js-translations-appear-not-to-be-covered-by-the-checks-see-analytics

Most of the testing was done in TC, where we confirmed it matches the current behaviour.

yob commented on lib/i18n/hygiene/tasks/check_variables.rb:
```diff
@@ -0,0 +1,41 @@
+        end
+      end
+    end
+  end
```
To use this task at the moment the user has to require this file in their rakefile. Should we require this file by default when the gem is required?

keithpitty replied:
Excellent suggestion.

nickbrowne replied:
I'd originally done it that way, but having looked at Rubocop and Cane, they do not do this for classes which can define rake tasks. I suspect this might be because they don't have a dependency on rake, so it's a way of allowing use of those gems without rake.

In our case I have added a dependency on rake as there's not a convenient interface to use it via plain old ruby.

So... yes!

nickbrowne replied:
It's now required by default

keithpitty commented:
> Most of the testing was done in TC, where we confirmed it matches the current behaviour.

Excellent. In that case, it's 👍 from me.

nullobject commented on spec/lib/i18n/hygiene/tasks/check_variables_spec.rb:
```diff
@@ -0,0 +1,18 @@
+require 'i18n/hygiene/tasks/check_variables'
+require 'i18n/hygiene/variable_checker'
+require 'i18n/hygiene/wrapper'
+require 'pry'
```
Do you need to require pry?

nickbrowne replied:
Ah whoops, that was supposed to be temporary!

nickbrowne replied:
Removed

nickbrowne committed:
Require CheckVariables

yob commented:
>> One thing worth noting is that I had to manually specify the :environment task as a dependency to ensure rails (and our locale data) was loaded.

> Can you think of a way to avoid this? It does seem like a bit of a gotcha.

Not easily.

If we were to make this a rails-only gem then we could at least raise a helpful error if the `Rails` constant wasn't available.

Would it make sense to make the tasks (in `lib/i18n/hygiene/tasks/check_variables.rb`) fail with a helpful message if `wrapper.keys_to_check` is empty?

yob commented:
I've tested this in donations and confirmed that it detects a variable mismatch in the `fr` locale and passes when that's fixed 👍

yob commented:
It would be nice if `I18n::Hygiene::VariableChecker` allowed the main locale to be configured too (it's currently hard-coded to `:en`), but it's not required for the gem to be useful across TC projects so I think we can add that to the wishlist.

nickbrowne commented:
> Would it make sense to make the tasks (in lib/i18n/hygiene/tasks/check_variables.rb) fail with a helpful message if wrapper.keys_to_check is empty?

We could, but what would that error message say? keys_to_check could be empty for a few different reasons.

nickbrowne commented:
> Not easily.
> 
> If we were to make this a rails-only gem then we could at least raise a helpful error if the Rails constant wasn't available.

I wonder if an alternative could be to inject a dependency on `:environment` `if defined?(Rails)`?

yob commented:
> I wonder if an alternative could be to inject a dependency on `:environment` `if defined?(Rails)?`

I'm completely unfamiliar with this style of building rake tasks, but if that's possible it might be worth trying.

I can't decide what I think - on the one hand it seems surprising to auto add a depency, on the other hand there's a high chance of users getting it wrong

nickbrowne commented:
Yup, though there's also the surprise of it not "just working" to consider. I'll have a play around and see if I can get something that isn't horrible.

nickbrowne committed:
Inject dependency on :environment if Rails is defined

This is mostly to make these rake tasks more convenient to use.

nickbrowne commented:
8304bb77 caused a bunch of spec failures (automated builds please!) dealing with it now.

nickbrowne commented:
Rebased to fixup a commit that introduced spec failures, also squashed the commit that removed the 'pry' require.

nickbrowne commented on lib/i18n/hygiene/tasks/check_variables.rb:
```diff
@@ -0,0 +1,47 @@
+            desc %(Check i18n hygiene)
+          end
+
+          task(task_name => dependencies) do
```
@yob I don't think this is too horrible
